docs: Improve netbase comments

dongcarl opened the pull request:
Second in a series of PRs documenting the net stack. Contributed with sincere thanks to sipa, laanwj, and gmaxwell for providing much of the history, context, and rationale.

dongcarl committed:
docs: Improve netbase comments

- Improve and add various Lookup* docs
- Improve InterruptibleRecv docs
- Improve Socks5 docs
- Add CreateSocket docs
- Add ConnectSocketDirectly docs
- Add SetNameProxy docs
- Add ConnectThroughProxy docs
- Add LookupSubNet docs

practicalswift commented:
Concept ACK

Very nice!

jonasschnelli commented:
Concept ACK
ping @theuni

instagibbs commented:
concept ACK!

ariard commented on src/netbase.cpp:
```diff
@@ -137,6 +166,26 @@ bool LookupHost(const char *pszName, CNetAddr& addr, bool fAllowLookup)
+/**
+ * Resolve a service string to its corresponding service.
+ *
+ * @param pszName The string representing a service. Could be a name or a
```
nit: maybe align name params and description comments to ease readability, at least some functions in others parts of code are doing it

dongcarl replied:
Not sure what exact alignment you mean, but my `netaddress` comments have been aligned like this as well. Lmk!

ariard replied:
Was thinking something likely to `BroadcastTransaction` in `transaction.h`, but that's a nit, doesn't matter!

ariard commented on src/netbase.cpp:
```diff
@@ -165,6 +220,16 @@ bool Lookup(const char *pszName, CService& addr, int portDefault, bool fAllowLoo
 }
 
+/**
+ * Resolve a service string with a numeric IP to its first corresponding
```
maybe add a TODO to refactored out LookupNumeric by first-corresponding-service Lookup it seems feasible?

laanwj replied:
please don't add TODOs to the comments, especially not for refactors

ariard commented on src/netbase.cpp:
```diff
@@ -234,21 +299,30 @@ enum class IntrRecvError {
+ *          IntrRecvError::OK only if all of the specified number of bytes were
+ *          read.
+ *
+ * @note This function requires that the specified socket is in non-blocking
```
redundancy between param hSocket and the note, maybe in param description stress on with "has to be in non-blocking mode"

dongcarl replied:
Done. Thanks!

ariard commented on src/netbase.cpp:
```diff
@@ -234,21 +299,30 @@ enum class IntrRecvError {
+ *
+ * @see This function can be interrupted by calling InterruptSocks5(bool).
+ *      Sockets can be made non-blocking with SetSocketNonBlocking(const
+ *      SOCKET&, bool).
```
nit: SetSocketNonBlocking(const SOCKET&, true), bool doesn't imply socket is blocking/non-blocking (and maybe add a TODO to refactor SetSocketNonBlocking without fNonBlocking flag, which seems useless given function name expected semantic)

dongcarl replied:
I wrote `SetSocketNonBlocking(const SOCKET&, bool)` so that Doxygen would render it and auto-link to the correct function if we ever made this function public.

ariard replied:
Ah okay, need to unset non-blocking socket was removed long time ago in #4869, but maybe there could be an usage of it for future new types of socket..

ariard commented on src/netbase.cpp:
```diff
@@ -234,21 +299,30 @@ enum class IntrRecvError {
     int64_t curTime = GetTimeMillis();
     int64_t endTime = curTime + timeout;
-    // Maximum time to wait in one select call. It will take up until this time (in millis)
+    // Maximum time to wait for I/O. It will take up until this time (in millis)
```
nit: maximum to wait for "readyness" of any I/O operation?

dongcarl replied:
Done. Thanks!

ariard commented on src/netbase.cpp:
```diff
@@ -68,6 +68,8 @@ bool static LookupIntern(const char *pszName, std::vector<CNetAddr>& vIP, unsign
 
     {
         CNetAddr addr;
+        // We can't actually resolve onion addresses, return a network address
```
^ above, nit: maybe ParseNetwork could get also a little comment on which networks are currently supported due to addr v1, notably diff between tor and onion

dongcarl replied:
Addressed in https://github.com/bitcoin/bitcoin/pull/16029

promag commented:
Nice, concept ACK.

dongcarl commented:
Addressed comments

ariard commented:
ACK 375d9c8

theuni commented:
Concept ACK. Will review.

sipa commented on src/netbase.cpp:
```diff
@@ -68,6 +68,8 @@ bool static LookupIntern(const char *pszName, std::vector<CNetAddr>& vIP, unsign
 
     {
         CNetAddr addr;
+        // We can't actually resolve onion addresses, return a network address
```
I think it would be clearer to point out that we treat ".onion" addresses the same way as we treat IPv4 a.b.c.d or IPv6 hex addresses, with the exception that the standard library resolving function don't support them.

dongcarl replied:
> I think it would be clearer to point out that we treat ".onion" addresses the same way as we treat IPv4 a.b.c.d or IPv6 hex addresses

Not 100% sure what you mean here, specifically, what is your specific meaning of "treat the same way"?

sipa replied:
For example "1.2.3.4" is a name, but it's one that directly encodes an IP address; no resolving is done when you turn it into a CNetAddr, it's just converted. The same is true for onion addresses; from our perspective, those are not hostnames - they're direct encodings of CNetAddr for onion addresses.

In contrast, "sipa.be" is a hostname that needs network activity to resolve to a CNetAddr.

dongcarl replied:
Updated: https://github.com/bitcoin/bitcoin/pull/15824/files#diff-4942713bcd2752bb531bf76c9a0ba38eR71

sipa commented on src/netbase.cpp:
```diff
@@ -234,22 +299,29 @@ enum class IntrRecvError {
- * Read bytes from socket. This will either read the full number of bytes requested
- * or return False on error or timeout.
- * This function can be interrupted by calling InterruptSocks5()
+ * Try to read a specified number of bytes from a socket. Please read the "note"
```
What note?

dongcarl replied:
Removed!

sipa reviewed (commented):
ACK

theuni commented on src/netbase.cpp:
```diff
@@ -501,12 +617,17 @@ bool ConnectSocketDirectly(const CService &addrConnect, const SOCKET& hSocket, i
         if (nErr == WSAEINPROGRESS || nErr == WSAEWOULDBLOCK || nErr == WSAEINVAL)
         {
+            // Connection didn't actually fail, but is being established
+            // asynchronously. Thus, use async I/O api (select/poll) to check
```
Nit: "async I/O api (select/poll) synchronously". Just so nobody goes looking for the missing magic to make this happen in the background.

dongcarl replied:
Changed.

theuni commented on src/netbase.cpp:
```diff
@@ -572,6 +702,20 @@ bool GetProxy(enum Network net, proxyType &proxyInfoOut) {
 }
 
+/**
+ * Set the proxy to use for populating our peer list using a one-shot `getaddr`
```
I believe most of this comment is incorrect, or at least very misleading :(.

SetNameProxy makes sure that all resolves are routed through a proxy rather than being done ourselves. No need to bring the higher-level 'getaddr' into it.

Maybe you grepped for HasNameProxy and missed the use of GetNameProxy in ConnectNode()? If so, looking at that may change your understanding somewhat.

dongcarl replied:
Okay this might also inform my documentation on `net.cpp` proper. I see that in the "Resolve" stage of `ConnectNode`, we are guaranteed a non-external query lookup if we have a name proxy. However, if `pszDest` was a hostname, the `Lookup` would fail, and we drop down to the `else if (pszDest && GetNameProxy(proxy))` case in the "Connect" stage, where we connect to the host through the proxy without doing any resolution ourselves (and delegate the resolution to the proxy). Is that a correct understanding?

sipa replied:
That matches what I remember about this logic. The reason is that if we're using Tor, we don't want to perform DNS lookups ourselves (as those would leak to DNS servers, instead of over the proxy). Instead, we do a name-based connection where we give the name directly as host to connect to the proxy (which in the case of Tor will cause the exit node to do the resolving locally and establish a connection with the result).

dongcarl replied:
Reworded here: https://github.com/bitcoin/bitcoin/pull/15824/files#diff-4942713bcd2752bb531bf76c9a0ba38eR714

theuni reviewed (commented):
This was much-needed, thank you!

Because this is documentation of some of our most quirky quirks, I was a little extra pedantic in my review.

theuni commented on src/netbase.cpp:
```diff
@@ -77,15 +79,21 @@ bool static LookupIntern(const char *pszName, std::vector<CNetAddr>& vIP, unsign
         return false;
 
+    // Traverse the linked list starting with aiTrav, add all non-internal
+    // IPv4,v6 addresses to vIP while respecting nMaxSolutions.
```
Maybe @EthanHeilman has some experience with this?

EthanHeilman replied:
I would have assumed just reading this code that the OS would not impose an order on the addresses returned. However after @theuni's comment I see no reason why that assumption is justified. It seems entirely plausible that the OS or some upstream source would impose an order. Randomizing the order of the linked list would negate all of this.

theuni replied:
@EthanHeilman Because AF_UNSPEC actually results in two separate/parallel resolves internally, I would assume that for performance reasons the final list would simply be those two lists concatenated.

EthanHeilman replied:
@theuni Even if it turns out the OS randomizes the list, that make not hold in the future and/or across OSes. If we expect the lists to be randomized in this code and it seems like we do, we should randomize the lists ourselves. Seems like a neat starter project.

dongcarl replied:
@theuni @EthanHeilman My naive thought is to just pop off the i-th item in the linked list each iteration and process it, where i = random_inclusive(0, num_remaining - 1). For most cases we're only interested in 1 item, so that should be pretty cheap. We would of course have to traverse the list first to find the initial length.

Since the list fits in memory I think reservoir sampling is overkill. But it does avoid having to do linked-list manipulation, initial traversal. However, it requires us to do random number generation for each item of the list.

Is there a better way you guys are thinking of?

EthanHeilman replied:
I was just thinking about it with regard to seeds. Not sure what other features randomization would be important for.

dongcarl replied:
@EthanHeilman It occurred to me that this doesn't really matter until we decrease the `nMaxIPs` when learning from seeds. Right now it's 256, and no peer returns more than that, so there's no need to shuffle as there's no cut-off.

dongcarl commented on src/netbase.cpp:
```diff
@@ -77,15 +79,21 @@ bool static LookupIntern(const char *pszName, std::vector<CNetAddr>& vIP, unsign
     aiHint.ai_protocol = IPPROTO_TCP;
+    // We don't care which address family (IPv4 or IPv6) is returned
     aiHint.ai_family = AF_UNSPEC;
+    // If we allow lengthy lookups of hostnames, only return addresses whose
```
What privacy concerns are there? What call-sites are you talking about?

sipa replied:
DNS lookups leak your query to your DNS server (or when done over tor, to the exit node that does the resolving).

dongcarl replied:
@sipa If I'm understanding correctly, one such instance is the "Resolve" stage of `ConnectNode` where if we have a name proxy, `Lookup` won't do external queries?

sipa replied:
If you set AI_NUMERICHOST then all `getaddrinfo` does is parse the IP address, and return the result immediately without network interaction.

dongcarl replied:
Ah, I thought that was obvious and wanted to explain why we needed `AI_ADDRCONFIG` for the case where `fAllowLookup` was true instead of just giving no flags. I should probably talk about both though.

dongcarl replied:
Updated: https://github.com/bitcoin/bitcoin/pull/15824/files#diff-4942713bcd2752bb531bf76c9a0ba38eR91

dongcarl commented on src/netbase.cpp:
```diff
@@ -137,6 +166,26 @@ bool LookupHost(const char *pszName, CNetAddr& addr, bool fAllowLookup)
+ * @param portDefault The default port for resulting services if not specified
+ *                    by the service string.
+ * @param fAllowLookup Whether or not we allow lengthy lookups by hostnames, or
+ *                     only allow parsing of numerical IPs.
```
Perhaps instead of "external queries", it's more accurate to say "hostname lookup"? I'm thinking about the case where it's an entry in the `/etc/hosts` file and that's not external.

dongcarl replied:
Updated: https://github.com/bitcoin/bitcoin/pull/15824/files#diff-4942713bcd2752bb531bf76c9a0ba38eR189

dongcarl commented on src/netbase.cpp:
```diff
@@ -464,17 +566,18 @@ SOCKET CreateSocket(const CService &addrConnect)
     SetSocketNoDelay(hSocket);
 
-    // Set to non-blocking
+    // Set the non-blocking option on the file descriptor.
```
Changed.

dongcarl commented:
Updated and addressed all concerns.

laanwj commented on src/netbase.cpp:
```diff
@@ -572,6 +710,21 @@ bool GetProxy(enum Network net, proxyType &proxyInfoOut) {
+ * @returns Whether or not the operation succeeded.
+ *
+ * @note Alternatively, SOCKS5 does have support for UDP-over-SOCKS5, but a DNS
+ *       resolver is beyond the scope of this project.
```
Note that no SOCK5 server in common use, most notably Tor, actually implements UDP support. So it'd be kind of pointless to do this.
The [Tor socks extension spec](https://github.com/torproject/torspec/blob/master/socks-extensions.txt) does mention a RESOLVE command that they implement that is specific to Tor, this would be a more realistic option.

dongcarl replied:
@laanwj Cool, will include. I'm thinking we use `@note` as a place to discuss the design space we've considered and why they don't work so it's recorded somewhere for future developers. Does that good to you? We could also do this somewhere else and reference from `@note`.

laanwj replied:
Yes, sounds good to me, I was just mentioning it to prevent people from going up this road to implement this (reading it as a TODO) only to conclude it's pointless, so it makes sense to mention that.

dongcarl replied:
Reworded a little bit to make it more clear this is just a design space comment not a TODO,

dongcarl commented:
Addressed all concerns and rebased.

Reviewers, you can verify that there were no substantial code changes that would invalidate my comments by invoking:

```sh
git log 65526fc8666fef35ef908dbc225f706bef642c7e..origin/master -- src/netbase.{h,cpp}
```

laanwj commented:
ACK c7f6ce74d3a5cf2a0c5bac20eab1efd997175a72
